function-specification.md

```diff
@@ -0,0 +1,83 @@
+# Function Specification
+
+Functions invoked by this framework are expected to conform to a well-defined
+contract. This document outlines the function signature that function developers
+must conform to as well as an outline for an API that function developers can
+use to receive and respond to events, subscribe to channels(?), and query the
+state of the system(?). The API and function signature are provided here in
+JavaScript, but consideration is given to the fact that there will be other
+function frameworks with different language constructs. (note: should we just
+do this in UML?)
+
+
+## Function Signature
+
+A function should accept some form of `Context` object. Depending on the language,
```

Comment:
I think you should also specify that the incoming cloud event and the outgoing cloud event are optional, so user must expect no cloud event as input and can do an empty reply.

In JS I assume optional means undefined, but in other languages (rust) is a well defined type `Option<Event>`

Reply:
I think there is a lot still we need to decide about cloud events, when they are used, when not. I suppose it makes sense that it would be optional as long as HTTP input/output is supported. We did talk earlier about the possibility of converting HTTP requests into events. But this eliminates support for streaming data, I believe.